ModFrmHilD/PositiveUnitReps.m

```diff
@@ -0,0 +1,677 @@
+      Append(~M`FunDomainRepsOfNorm[bb][Norm(nn)], nu);
+    else
+      M`FunDomainRepsOfNorm[bb][Norm(nn)] := [nu];
+    end if;
+  end for;
+
+  for x in [1 .. M`Precision] do
+    for bb in NarrowClassGroupReps(M) do
+      // It's already defined if there were some nus associated to nn of this norm
+      if IsDefined(M`FunDomainRepsOfNorm[bb], x) then
+        M`FunDomainRepsUpToNorm[bb][x] := M`FunDomainRepsUpToNorm[bb][x-1] cat M`FunDomainRepsOfNorm[bb][x];
+      else
+        M`FunDomainRepsUpToNorm[bb][x] := M`FunDomainRepsUpToNorm[bb][x-1];
+        M`FunDomainRepsOfNorm[bb][x] := [];
+      end if;
+    end for;
+  end for;
+end intrinsic;
+
+intrinsic ComputeShadows(M::ModFrmHilDGRng, bb::RngOrdFracIdl) -> Assoc
+  {
+    inputs:
+      M: A graded ring of Hilbert modular forms
+      bb: Fractional ideal of the ring of integers of the number field underlying M
+    returns: 
+      An associative array shadows_bb keyed by norm x whose values at x is an enumerated set
+      storing pairs (eps, nu), where eps is a totally positive unit and nu
+      a fundamental domain representative, such that the element nu*eps is dominated by 
+      (totally less than) some fundamental domain representative nu' with norm at most x. 
+
+      Practically, the elements eps*nu coming from shadows_bb[x] are the extra indices
+      which must be accounted for when multiplying ModFrmHilDEltComps of precision x.
+
+
+    Let log_max_coord be the logarithm of the largest coordinate entry appearing
+    in any representative nu. The algorithm enumerates totally positive elements
+    of the field lying in the intersection of the n-dimensional hypercube with 
+    opposite corners (0, .., 0) and (log_max_coord, .., log_max_coord) with the 
+    region of the totally positive orthant with norm at most M`Precision. 
+
+    For each such point (which we call a candidate shadow), it then checks 
+    if it is dominated by some representative. If it is, then we call it a shadow
+    and add it to our output. 
+
+    TODO abhijitm Needing to check domination after the fact is an annoying requirement
+    arising because this hypercube intersect small norm region is generally bigger than 
+    the correct region. I think I know of a better search algorithm but it'd be a lot of 
+    work so I feel this is good enough for now. 
+  }
+
+  RR_PREC := 100;
+  FUZZ := 10^-29;
+  RR := RealField(RR_PREC);
+
+  F := BaseField(M);
+  epses := TotallyPositiveUnitsGenerators(F);
+  ZF := Integers(F);
+  n := Degree(F);
+  places := InfinitePlaces(F);
+  bbp := NarrowClassGroupRepsToIdealDual(M)[bb];
+
+  reps := FunDomainRepsUpToNorm(M)[bb][M`Precision];
+  nu_norms := {Norm(nu) : nu in reps | not IsZero(nu)}; 
+  min_norm := Min(nu_norms);
+
+  // The largest coordinate entry appearing in any representative nu.
+  // This determines the hypercube we work with
+  max_coord := Max(&cat[[Evaluate(nu, v) : v in places] : nu in reps]) + FUZZ;
```

Comment:
Yeah, enough other parts of the code are broken for non-Galois fields anyways that I decided it was okay.

Reply:
I think if we really want to optimize this someday we should compute shadows "properly" instead of doing this superset then sieve approach.

Reply:
It's true that we could compute this polytope! It would have many edges though. And my impression is that as the precision grows the part we have to sieve out is like asymptotically negligible, but I might be wrong.